### ISSUE TITLE
LabelSampler Memory Consumption

### ISSUE THREAD

**Issue by meghbhalerao** [fepegar/torchio#397]
Hi All,
I am using Label Sampler as follows:
```python
psize = (144,144,80)
sampler = torchio.data.LabelSampler(psize)
patches_queue = torchio.Queue(subjects_dataset, max_length=1,
                                samples_per_volume=1,
                                sampler=sampler, num_workers=4,
                                shuffle_subjects=False, shuffle_patches=True, verbose=False)
```
The `LabelSampler` seems to consume CPU memory upwards of **120GB**. I have 160ish training images. I am using about _6 augmentations_ with `probability = 0.1`. 

However, when I use the exact same configuration, but with `UniformSampler`, the CPU memory requirement is much lesser.

Any reason this is happening?
Thanks for your help,
Megh

**Comment by fepegar** [comment]
I don't know what could be happening. Can you try the `WeightedSampler` using your label map name as probability map name? The result should be the same, if you have only one class in your label maps.

Do you think you can provide a minimally working example? Maybe using `torchio.utils.create_dummy_dataset`.

**Comment by meghbhalerao** [comment]
Thank you @fepegar. Sure, I will try using `WeightedSampler` and let you know. Will try making a dummy dataset and let you know.
Thanks again

**Comment by meghbhalerao** [comment]
Hi @fepegar, I am currently training with `WeightedSampler` - this does not seem to be taking too much memory. 

Can I share 3 images from my dataset? Here is the link: https://drive.google.com/file/d/1nFkUIEJooyeGevtK0Y6qz022tMk-10Qo/view?usp=sharing

Since I was unsure how to use the dummy dataset exactly, I will also share the code to make the dataloader with you (it is like a pseudocode), here is the link. Hope this is okay: https://gist.github.com/meghbhalerao/84ed0c3db4db43f77bf4e5793c4debd4

Here are the augmentations with the corresponding probabilities that I am using:
```python
data_augmentation: 
  {
  'spatial':{
    'probability': 0.1
  },
  'kspace':{
    'probability': 0.1
  },
  'bias':{'probability': 0.1},
  'blur':{'probability': 0.1},
  'noise':{'probability': 0.1},
  'swap':{'probability': 0.1}
  }
```
Please let me know if something is not clear.
Thanks,
Megh

**Comment by fepegar** [comment]
Thanks for sharing! I don't see anything suspicious, I'm not sure how I can help. The LabelSampler is pretty much a WeightedSampler, so I don't know what could be wrong. I can maybe advise about that code, though.

1. You can use `ScalarImage` and `LabelMap` instead of `Image`
2. `RandomMotion` is slower than the other k-space transforms, so you might want to give it a smaller probability in `OneOf`, to go faster.
3. Why are you using `RandomSwap`? I don't think it's very helpful except in a self-supervised scenario...
4. You might want to decrease a bit the std of your `RandomBlur`. Using 4 for your images has this effect:

![Screenshot](https://user-images.githubusercontent.com/12688084/103207165-9346a500-48f5-11eb-9875-f91e6bb497db.png)

Remember you can use the [CLI or Slicer](https://torchio.readthedocs.io/interfaces/index.html) to preview the effect of the transforms.

**Comment by meghbhalerao** [comment]
Thanks a lot @fepegar for your detailed analysis. 

I observed one thing. When I am doing padding for `LabelSampler` and `WeightedSampler` as per our discussion here: https://github.com/fepegar/torchio/issues/392 - like this:
```python
psize = (144,144,80)
psize_pad = list(np.asarray(np.round(np.divide(psize,2)), dtype=int))
padder = Pad(psize_pad, padding_mode = 'symmetric') # for modes: https://numpy.org/doc/stable/reference/generated/numpy.pad.html
subject = padder(subject)
```
By doing this **padding** I observed that the **memory requirement** shoots up by a lot (>125GB). I am trying to investigate what the issue might be.
Thanks again, 
Megh

**Comment by sarthakpati** [comment]
FYI, even I am seeing similar shooting memory consumption for other datasets when performing padding. Unsure why this would be the case, because when I do padding manually (using ITK), it isn't that high.

**Comment by fepegar** [comment]
RAM usage depends on many things: images per subject, size of your images, transforms used, queue length... It's hard to know without data and code. For example, if you use the padding technique in #392 with a 512x512x256 image and a patch size of 128x128x128 and you have a T1, T2, Flair, T1 GAD and Segmentation, all float32, that's 5 * (512 + 128) * (512 + 128) * (512 + 128) * 4 / 2**30 = 4.88 GB. That's just the size of the loaded image. If you add computationally intensive transforms to that, it will be tough. If you use a queue length of 100, that's 3.9 GB more.

**Comment by sarthakpati** [comment]
So, I was trying to do this without any queue/augmentations - just the padding.

**Comment by fepegar** [comment]
Closing this for now. Please let us know if you find out anything new.
